### PR TITLE
SWE22 Template Editor

### LINKED ISSUE #7: Instructors Create Seating Templates for Each Course
As an instructor, I need to create a seating template for each course so that I can post it for students to select their seats,

Instructors will use a master template to activate seats as:
-open (anyone can sit there)
-closed (no one can sit there)
-reservable (registered students can reserve)
All other seats in the template will be removed.

The system doesn't need to offer instructors information about what they need

### LINKED ISSUE #22: Template Editor
Includes radio buttons to either make grid squares Open, Closed, Reservable. 
Needs save button

### PR DESCRIPTION
**Pull request by jsmith**
This PR resolves #22 
This PR resolves #7 

This PR implements the template editor. An instructor can now edit their template and set seats available / closed / reserved.

To make this all work, I added several RESTful endpoints. The full list of new endpoints includes:
- GET `/instructor/{courseName}/{courseSection}`: This is the view page where an instructor can edit their course template
- GET `/courses`: Retrieves all of the courses
- POST `/courses`: Insert or update a course
- POST `/courses/{id}/seats`: Insert or delete a particular course's seats
- GET `/course/{id}`: Retrieve a particular course

A few things that still need to be fixed are:
- [x] Ability to save changes to the database
- [x] Some more styling
- [x] Change view url to something like this: `intructor/{courseName}/{section}`

### CODE REVIEW

**Comment by jsmith** [comment]
Ok an instructor can now edit their templates! I'll add some comments throughout the PR

File: src/main/java/com/unb/tracker/Course.java [hunk 1]
```diff
@@ -1,23 +1,48 @@
...
+import javax.persistence.*;
+import java.util.List;
 
 @Entity // This tells Hibernate to make a table out of this class
 public class Course {
     @Id
     @GeneratedValue(strategy=GenerationType.AUTO)
-    private Integer id;
-
+    private Long id;
```
Comment (jsmith) [inline src/main/java/com/unb/tracker/Course.java:10]: Had to change this to a long for the findOne method!

File: src/main/java/com/unb/tracker/TrackerController.java [hunk 2]
```diff
@@ -18,22 +21,22 @@ public String index(ModelMap map) {
 
     @Autowired
```
Comment (jacsmith21) [inline src/main/java/com/unb/tracker/TrackerController.java]: Moved both autowired repositories to the top of the file as they can be used by any method

File: src/main/java/com/unb/tracker/TrackerController.java [hunk 3]
```diff
@@ -57,4 +60,37 @@ public String instructor(ModelMap map) {
         // This returns a JSON or XML with the users
         return userRepository.findAll();
     }
+
```
Comment (jacsmith21) [inline src/main/java/com/unb/tracker/TrackerController.java]: The following method includes the actual page where instructors can edit their template, the rest are RESTful methods

File: src/main/resources/static/js/grid.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    this.states = opt.states;
+    this.parent = opt.parent;
+    this.state = 0;
+
+    this.setState = function(state) {
+        this.el.removeClass(this.states[this.state]);
+        this.state = state;
+        this.el.addClass(this.states[this.state]);
+    };
+
```
Comment (jacsmith21) [inline src/main/resources/static/js/grid.js]: The following on click function implements the state functionality. Currently, I removed the radio buttons and just cycle through the various states. This was the easiest to implement and we can change it later if necessary

File: src/main/resources/templates/instructor/base_instructor.twig [hunk 5]
```diff
@@ -0,0 +1,25 @@
+{% extends '/templates/base.twig' %}
```
Comment (jsmith) [inline src/main/resources/templates/instructor/base_instructor.twig:1]: I created a base instructor class which inherits from base.twig to minimize repeated html

File: src/main/resources/templates/instructor/course.twig [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        </div>
+    </div>
+{% endblock %}
+
+{% block script %}
+    {{ parent() }}
+    <script type="text/javascript" src="/js/grid.js"></script>
+    <script>
+        $(document).ready(function() {
+            var grid;
```
Comment (jacsmith21) [inline src/main/resources/templates/instructor/course.twig]: This ajax call runs initially when the page loads and then renders the grid

File: src/main/java/com/unb/tracker/TrackerController.java [hunk 7]
```diff
@@ -57,4 +60,37 @@ public String instructor(ModelMap map) {
...
+    @GetMapping(value="/instructor/{courseId}") // courseId should be changed to name / section eventually
+    public String getCoursePage(@PathVariable Long courseId, ModelMap map) {
```
Comment (jsmith) [inline src/main/java/com/unb/tracker/TrackerController.java]: The following method includes the actual page where instructors can edit their template, the rest are RESTful methods

File: src/main/resources/static/js/grid.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    this.el.on('click', {cell: this}, function(e) {
+        cell = e.data.cell;
```
Comment (jsmith) [inline src/main/resources/static/js/grid.js:18]: The following on click function implements the state functionality. Currently, I removed the radio buttons and just cycle through the various states. This was the easiest to implement and we can change it later if necessary

File: src/main/resources/templates/instructor/course.twig [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            $.ajax({
+                type: "get",
+                url: "/courses/{{ course.id }}",
```
Comment (jsmith) [inline src/main/resources/templates/instructor/course.twig:36]: This ajax call runs initially when the page loads and then renders the grid

File: src/main/resources/templates/instructor/course.twig [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+            $('#editsave').on('click', function(e) {
+                if(grid.editable) {
+                    grid.editable = false;
+                    $('#editsave').html('Edit');
+                    var seats = grid.getSeats();
+                    console.log(seats);
+                    $.ajax({
+                        type: "post",
+                        url: "/courses/{{ course.id }}/seats",
```
Comment (jsmith) [inline src/main/resources/templates/instructor/course.twig:60]: This ajax call runs when the save button is clicked to save the changes in the database

**Comment by brandon1024** [comment]
Hey @jacsmith21 , in the issue description, can you provide step by step instructions on how to view/test the changes you've made? For some reason, after I create a course and go to http://localhost:8080/courses/{course_id_here}, im just getting a JSON: `{"id":1,"name":"test","rows":5,"cols":5,"seats":[]}`. No grid view, and I'm not sure what I'm doing wrong :p

**Comment by jsmith** [comment]
Oh yeah sorry about that, I'll update the PR description. The reason that was happening is that I created a few RESTful endpoints for courses!

**Comment by brandon1024** [comment]
Ok now I can get to the grid view. Sweet. I do see one issue though;

When I visit `http://localhost:8080/instructor/1` and go to click on a link in the course list, I get a 404 because it's trying to navigate to `http://localhost:8080/instructor/courses/1`

**Comment by jsmith** [comment]
Oh nice, I'll fix that bug!

**Comment by jsmith** [comment]
@brandon1024 The list of TODOs has been finished!

**Comment by brandon1024** [comment]
@jacsmith21 Reviewing

**Comment by brandon1024** [comment]
@jacsmith21 Looks great! The only issue I see is that when you are in `http://localhost:8080/instructor/coursename`, theres no way to get back to the instructor view. Maybe make the navbar `Attendance Tracker` text a link to `http://localhost:8080/instructor/`? Once this is working, good to merge!

**Comment by jsmith** [comment]
Added those changes 👍

**Review by brandon1024 — APPROVED** [review]
Looks fab! Merging

**Pull request merged**